src/superstream_producer.rs

```diff
@@ -0,0 +1,148 @@
+use crate::error::ProducerCloseError;
+use crate::{
+    client::Client,
+    environment::Environment,
+    error::{ProducerCreateError, ProducerPublishError},
+    producer::{ConfirmationStatus, NoDedup, Producer},
+    superstream::{DefaultSuperStreamMetadata, RoutingStrategy},
+};
+use rabbitmq_stream_protocol::message::Message;
+use std::collections::HashMap;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+type FilterValueExtractor = Arc<dyn Fn(&Message) -> String + 'static + Send + Sync>;
+
+#[derive(Clone)]
+pub struct SuperStreamProducer<T>(
+    Arc<SuperStreamProducerInternal>,
+    HashMap<String, Producer<T>>,
+    DefaultSuperStreamMetadata,
+    PhantomData<T>,
+    RoutingStrategy,
+);
+
+/// Builder for [`SuperStreamProducer`]
+pub struct SuperStreamProducerBuilder<T> {
+    pub(crate) environment: Environment,
+    pub filter_value_extractor: Option<FilterValueExtractor>,
+    pub route_strategy: RoutingStrategy,
+    pub(crate) data: PhantomData<T>,
+}
+
+pub struct SuperStreamProducerInternal {
+    pub(crate) environment: Environment,
+    client: Client,
+    // TODO: implement filtering for superstream
+    filter_value_extractor: Option<FilterValueExtractor>,
+    //routing_strategy: RoutingStrategy,
+}
+
+impl SuperStreamProducer<NoDedup> {
+    pub async fn send<Fut>(
+        &mut self,
+        message: Message,
+        cb: impl Fn(Result<ConfirmationStatus, ProducerPublishError>) -> Fut
+            + Send
+            + Sync
+            + 'static
+            + Clone,
+    ) -> Result<(), ProducerPublishError>
+    where
+        Fut: Future<Output = ()> + Send + Sync + 'static,
+    {
+        let routes = match &self.4 {
+            RoutingStrategy::HashRoutingStrategy(routing_strategy) => {
+                routing_strategy.routes(&message, &mut self.2).await
+            }
+            RoutingStrategy::RoutingKeyStrategy(routing_strategy) => {
+                routing_strategy.routes(&message, &mut self.2).await
+            }
+        };
+
+        for route in routes.into_iter() {
+            if !self.1.contains_key(route.as_str()) {
+                let producer = self
+                    .0
+                    .environment
+                    .producer()
```

Comment:
This one producer() is returning as error ProduceCreateError while the producer.send and super_stream_prdoducer_send() is returning ProducerPublishError. We should find a way to manage this error type difference

Reply:
As discussed created SuperStreamProducerPublishError